### PR TITLE
Efficiency should be 0 when there is no path between nodes

### PR DESCRIPTION
**Pull request by arodriguezca**
Currently, when there is no path between nodes, it throws NetworkXNoPath. However, the [author](http://www.uvm.edu/pdodds/files/papers/others/2001/latora2001a.pdf) said it should be 0. Refer to page 2, the paragraph right above formula 1.

### CODE REVIEW

**Comment by dschult** [comment]
This looks good. 
Could you adjust the docstring so it is fewer than 80 chars wide? (pep8 style).
Thanks!

**Comment by dschult** [comment]
If you are up for it, adding a test to tests/test_efficiency.py would be great!

**Comment by arodriguezca** [comment]
Style fixed. @dschult, I can do the test. I'll code it later this week.

**Comment by jarrodmillman** [comment]
@arodriguezca It would be great if you could add a test in the next few days.  If you have any questions, please let us know.

**Comment by jarrodmillman** [comment]
@arodriguezca I would like to merge this before we make a 2.0beta release later this week.  If you prefer, I can add a test for this.  Thanks!

**Comment by arodriguezca** [comment]
@jarrodmillman I was disconnected for some days, sorry for the inconveniences. I'll have the test commited by Wednesday night.

**Comment by jarrodmillman** [comment]
Great, thanks!

**Comment by arodriguezca** [comment]
@dschult @jarrodmillman besides the test for efficiency, I added the case of the global efficiency being 0 when the graph is disconnected (no edges) as it should be according to the paper.

**Comment by dschult** [comment]
These tests look good. But they should go in the ```networkx/algorithms/tests/test_efficiency.py``` file rather than the ```networkx/tests/``` directory.   Don't worry about the Codecov errors... those are fixed in another PR.

File: networkx/tests/test_efficiency.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        self.G1 = nx.barbell_graph(10, 3)
+        # G2 is a disconnected graph
+        self.G2 = nx.Graph()
+        self.G2.add_nodes_from([1, 2, 3])
+
+    def test_efficiency(self):
+        """
+        Returns efficiency of two nodes
+        """
+        return nx.efficiency(self.G1, 1, 10)
```
Comment (jarrodmillman) [inline networkx/tests/test_efficiency.py]: assert_equal(nx.efficiency(self.G1, 1, 10), 0.5)

File: networkx/tests/test_efficiency.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+    def test_efficiency_disconnected_nodes(self):
+        """
+        Returns 0 when nodes are disconnected
+        """
+        return nx.efficiency(self.G2, 1, 2)
```
Comment (jarrodmillman) [inline networkx/tests/test_efficiency.py]: assert_equal(nx.efficiency(self.G2, 1, 2), 0)

File: networkx/tests/test_efficiency.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    def test_local_efficiency(self):
+        """
+        Test local efficiency
+        """
+        return nx.local_efficiency(self.G1)
```
Comment (jarrodmillman) [inline networkx/tests/test_efficiency.py]: assert_almost_equal(nx.local_efficiency(self.G1), 0.971146245)

File: networkx/tests/test_efficiency.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    def test_local_efficiency_disconnected_graph(self):
+        """
+        In a disconnected graph the efficiency is 0
+        """
+        return nx.local_efficiency(self.G2)
```
Comment (jarrodmillman) [inline networkx/tests/test_efficiency.py]: assert_equal(nx.local_efficiency(self.G2), 0)

File: networkx/tests/test_efficiency.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+    def test_global_efficiency(self):
+        """
+        test global efficiency
+        """
+        return nx.global_efficiency(self.G1)
```
Comment (jarrodmillman) [inline networkx/tests/test_efficiency.py]: assert_almost_equal(nx.global_efficiency(G1), 0.525757575)

File: networkx/tests/test_efficiency.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
+
```
Comment (jarrodmillman) [inline networkx/tests/test_efficiency.py]: from nose.tools import assert_equal, assert_almost_equal

**Comment by arodriguezca** [comment]
@dschult I didn't notice that there was already one test file there. I merged them.

File: networkx/algorithms/tests/test_efficiency.py [hunk 7]
```diff
@@ -1,54 +1,80 @@
...
 #
-# Copyright 2015 NetworkX developers.
+# Copyright 2017 NetworkX developers.
 #
 # This file is part of NetworkX.
 #
 # NetworkX is distributed under a BSD license; see LICENSE.txt for more
 # information.
-"""Unit tests for the :mod:`networkx.algorithms.efficiency` module."""
+# Unit tests for the :mod:`networkx.algorithms.efficiency` module.
```
Comment (jarrodmillman) [inline networkx/algorithms/tests/test_efficiency.py]: This should remain a docstring.

File: networkx/algorithms/tests/test_efficiency.py [hunk 8]
```diff
@@ -1,54 +1,80 @@
 # test_efficiency.py - unit tests for the efficiency module
 #
-# Copyright 2015 NetworkX developers.
+# Copyright 2017 NetworkX developers.
```
Comment (jarrodmillman) [inline networkx/algorithms/tests/test_efficiency.py]: please use ``2015-2017``

File: networkx/algorithms/tests/test_efficiency.py [hunk 9]
```diff
@@ -1,54 +1,80 @@
...
-    """Tests that the average global efficiency of the complete graph is
-    one.
+    def test_local_efficiency(self):
+        """
+        Test local efficiency
+        """
+        return nx.local_efficiency(self.G1)
 
-    """
-    for n in range(10):
```
Comment (jarrodmillman) [inline networkx/algorithms/tests/test_efficiency.py]: It looks like the test got dropped.

Comment (jarrodmillman) [inline networkx/algorithms/tests/test_efficiency.py]: hmm... It also look it had a bug, which is why you probably didn't keep the for loop.

Why not put it back, but use:

```
for n in range(2, 10):
    G = nx.complete_graph(n)
    assert_equal(nx.global_efficiency(G), 1)
```

File: networkx/algorithms/tests/test_efficiency.py [hunk 10]
```diff
@@ -1,54 +1,80 @@
...
+        """
+        return nx.global_efficiency(self.G1)
 
-class TestLocalEfficiency(TestCase):
-    """Unit tests for the local efficiency function."""
+    def test_efficiency_cycle_graph(self):
+        assert_equal(nx.efficiency(self.G3, 0, 1), 1)
+        assert_equal(nx.efficiency(self.G3, 0, 2), 1 / 2)
 
-    def test_complete_graph(self):
```
Comment (jarrodmillman) [inline networkx/algorithms/tests/test_efficiency.py]: You could also do the same thing as I suggested for ``global_efficiency``:
```
for n in range(2, 10):
     G = nx.complete_graph(n)
     assert_equal(nx.local_efficiency(G), 1)
```

File: networkx/algorithms/tests/test_efficiency.py [hunk 11]
```diff
@@ -1,54 +1,80 @@
...
+    def test_using_ego_graph(self):
+        """
+        Test that the ego graph is used when computing local efficiency.
         For more information, see GitHub issue #2233.
-
         """
-        # This is the triangle graph with one additional edge.
-        G = nx.lollipop_graph(3, 1)
-        assert_equal(nx.local_efficiency(G), 23 / 24)
+        assert_equal(nx.local_efficiency(self.G5), 23 / 24)
```
Comment (jarrodmillman) [inline networkx/algorithms/tests/test_efficiency.py]: This should probably be:
```
assert_equal(nx.local_efficiency(self.G4), 23 / 24)
```

**Comment by arodriguezca** [comment]
@jarrodmillman Thanks for all your helpful comments! I fixed the things you pointed out and deleted some functions that I added because they were redundant.

**Pull request merged**